### ISSUE TITLE
Support for Anaconda `conda` environments from nushell

### ISSUE THREAD

**Issue by 00krishna** [nushell/nushell#2272]
**Is your feature request related to a problem? Please describe.**
I love nushell, but it seems that `nushell` is not currently compatible with the python Anaconda distribution. I was not sure where to direct this request, but I thought I would start with the nushell folks. So Anaconda is a python distribution for scientific computing. Users can create conda virtual environments and then activate those environments using the command `conda activate <environment_name>`.

When I tried to use `conda activate myenv` I got a message:

```
CommandNotFoundError: Your shell has not been properly configured to use 'conda activate'.
To initialize your shell, run

    $ conda init <SHELL_NAME>

Currently supported shells are:
  - bash
  - fish
  - tcsh
  - xonsh
  - zsh
  - powershell

See 'conda init --help' for more information and options.

```
Seems like this might be an upstream Anaconda issue, but I though I might check with the nushell team on whether there have been discussions about this. 

**Describe the solution you'd like**
I would like to be able to activate Anaconda environments using nushell.

**Describe alternatives you've considered**
The alternative is to stay with zsh for anything involving Anaconda or python. 

**Additional context**
No other info. Thanks for your consideration. I will post to the `conda` maintainers as well.

**Comment by fdncred** [comment]
I wonder if this is as simple as ensuring that the `activate.py` and other commands are in the path?

The exception is a `CommandNotFoundError`. You can see that [here](https://github.com/conda/conda/blob/568a948573a349cfa30b2fbb41ffc56014236887/conda/exceptions.py).

**Comment by 00krishna** [comment]
@fdncred thanks for the response. Yes, I use Ubuntu Linux exclusively, so `bat` files are not an issue thank goodness. So how would I try this from `nushell`? Like when you say that I need `activate.py` in the path, I am not sure which path you are referring to. Meaning that the `anaconda` folder is already in the `PATH` environment variable. Would I just have to start `nushell` and then `python activate.py`, or something? I don't mind being the guinea pig, but just need a sense of where to star on this.

**Comment by fdncred** [comment]
@00krishna Oh, I'm sorry. I thought this was in Windows.

When in nushell you should be able to do `echo $nu.path` and see what folders are in your path.

If you need to update your path in nushell you need to edit your config.toml file as this [page](https://www.nushell.sh/book/en/configuration.html) describes.

If your path in nushell already has the proper folder to access these files, then it's probably a bug in nushell. I'd start by trying regular conda commands like `conda env list` to see if it says you have environments. I'm not as familiar with linux so I'm not sure what happens when you type `activate blah`. I guess it searches the path for anything +x named activate but at some point, somewhere it has to look for activate.py unless there's an activate script somewhere.

Sorry I'm not much help.

**Comment by 00krishna** [comment]
@fdncred No worries. This is actually very helpful, it gets me started. I will try out what you suggest and see what happens. But thanks for chiming in.

**Comment by 00krishna** [comment]
@PranayAnchuri oh interesting. I use `zsh` too. I did not realize that the environment remains active inside of zsh, so that is nice to know. I have not had a chance to try out what @fdncred suggested, but I will give that a shot. I posted this to the `conda` team too and I think it got marked as a feature request. If I can get this working, as @fdncred suggested, then we can totally integrate this into the instructions or such. Seems like nushell and Anaconda are a natural fit.

**Comment by rfilmyer** [comment]
@00krishna do you have a link to that feature request on `conda` 's side, if it's being publicly tracked?

**Comment by filaretov** [comment]
Hey, I'm also interested in this.

conda/conda#10113 is @00krishna's issue on the conda repo.

As far as I can tell, there are a few things required to get this working.

1. conda environments are activated while the shell is running, which requires modifying the PATH. If I understand correctly, this is currently not supported in Nu. #3435 is the only relevant issue I found. [The entry on Bash equivalents in Nu](https://www.nushell.sh/book/coming_from_bash.html) also suggests that—there's no equivalent command for temporarily updating the path like `export PATH=...`. Is editing the `config.toml` the only way to change Nu's path?
2. Each shell has an activator class associated with it. [Here](https://github.com/conda/conda/blob/248741a843e8ce9283fa94e6e4ec9c2fafeb76fd/conda/activate.py#L995) is the one for fish. The most important bits seem to be a way to set and unset temporary variables (including the path), exporting values (though that's not really a concept in Nu, is it?) and sourcing an activation script, which brings us to:
3. An activation script. This seems to mostly handle changing the prompt and adding a wrapper around the conda command, which slightly adapts it based on which shell is being used. Once again, [here's an example](https://github.com/conda/conda/blob/master/conda/shell/etc/fish/conf.d/conda.fish) with fish. Similar to the path, the prompt has to be changed on the fly (and restored when deactivating a conda environment.

There a few other places in conda that have to be updated to support Nu, but I believe these are the most significant ones. I used fish as a reference because it's also a non-POSIX shell. I only read the parts in conda that referenced fish, so I might well have missed something.

It seems to me that the only thing we're missing in Nu to implement this is a way to change the path while the shell is running.

**Comment by filaretov** [comment]
Here's a very hacky workaround that might work while this is solved properly:

```
# in ~/.config/nu/config.toml
...
startup = ["""
def "conda activate" [env] {
  let nu-config-path = $"/home/($nu.env.USER)/.config/nu/config.toml"
  let config = (open $nu-config-path)
  let new-path = ($config.path | prepend $"/home/($nu.env.USER)/.conda/envs/($env)/bin")
  let new-env = (
  	$config.env
	| insert CONDA_DEFAULT_ENV $env
	| insert CONDA_PREFIX $"/home/($nu.env.USER)/.conda/envs/($env)"
	| insert CONDA_PROMPT_MODIFIER $"[($env)]"
	| insert CONDA_SHLVL "1"
	)
  let temp-conf = $"(mktemp -d | tr -d '\n')/config.toml"
  let new-config = ($config | update path $new-path | update env $new-env)
  $new-config | save $temp-conf
  nu --config-file $temp-conf
}
""",
"""
def "conda deactivate" [] {
    exit
}
""",
]
...
```

I've added these two subcommands to the `startup` section of my Nu `config.toml`

The `conda activate` subcommand adds the conda env bin path to `path` and a few conda environment variables and creates a new `config.toml` file for Nu in a temporary location. It assumes your conda environments are in `home/$user/.conda`, change that to whatever you need (I didn't have an environment variable that points to that directory, even though I'm pretty sure there should be one). It then spawns a new Nu instance using that new `config.toml`.

The `conda deactivate` command exits the shell. It would be better if it checks whether there's an activated conda environment before exiting.

All other conda commands run the conda executable itself (including its subcommands). This is made possible by Nu's great subcommand syntax, so thanks to the team for that!

There are probably some subcommands that won't run correctly like this. There are probably other things that won't work correctly too.

This depends on `mktemp` to create a temporary directory. This probably means this won't directly run on Windows or other systems that don't have that executable. If you have any suggestions for how to work around that, please share. I realize `with-env` or its variants would probably be a much better way to update these directly with Nu functions, but I didn't find a way to use the same approach for the path. If there's a way to do that, please let me know!

It seems to work well enough for my needs for now.

**Comment by fdncred** [comment]
@filaretov would you mind adding this script to the nushell/nu_scripts repo? If you can, just make a folder called python and put a readme.md and your activate and deactivate scripts in there. Feel free to reference this issue and just duplicate your text above.

**Comment by elferherrera** [comment]
@filaretov do you think you could sketch a similar script but only for `venv`. The simple version for activating an environment

**Comment by fdncred** [comment]
@filaretov Just throwing this out there too. We have a relatively new command called `load-env` that may be helpful for this type of thing.
```
~/source/repos/forks/nushell(main)                                                                                                                                        12|01:29:15.325 PM
└─ $ ▶ load-env --help
Set environment variables using a table stream

Usage:
  > load-env (environ) {flags}

Parameters:
  (environ) Optional environment table to load in. If not provided, will use the table provided on the input stream

Flags:
  -h, --help: Display this help message

Examples:
  Load variables from an input stream
  > echo [[name, value]; ["NAME", "JT"] ["AGE", "UNKNOWN"]] | load-env; echo $nu.env.NAME

  Load variables from an argument
  > load-env [[name, value]; ["NAME", "JT"] ["AGE", "UNKNOWN"]]; echo $nu.env.NAME

  Load variables from an argument and an input stream
  > echo [[name, value]; ["NAME", "JT"]] | load-env [[name, value]; ["VALUE", "FOO"]]; echo $nu.env.NAME $nu.env.VALUE
```

**Comment by filaretov** [comment]
@fdncred I'll clean it up and submit a pull request to nu_scripts.

@elferherrera Sure, I can do that!

I just have a question before I proceed. I noticed the relatively new `pathvar` command (#3828 and #3670). It seems like that's exactly what we need to implement proper conda and venv support. Is there any reason I shouldn't use that instead?

**Comment by fdncred** [comment]
>  Is there any reason I shouldn't use that instead?

if you think that's best, then please use it.

ya, `;` and `:` were swapped for a bit. sorry.

**Comment by elferherrera** [comment]
@filaretov thanks a lot. It will be easy if we start with that script and then jump to the larger conda.

Btw, you dont need to put the file in your config file. You can create a file with `load-env` and use `source` to create the env variables in your current shell

One thing that I could do was to create a deactivate function because the new function wont have access to the new env variables. You can try with a secondary script that deactivates the env variables (`unlet-env`)

**Comment by filaretov** [comment]
Hey, just an update: I have a working version that does not create a new shell [over here](https://github.com/filaretov/nu_scripts/blob/virtual-environments/virtual_environments/conda.nu)

In the same folder, you'll find one for Python venvs as well.

Both work by creating a table of env vars that the user has to load like this: `load-env (conda-env name-of-env)`.

They are deactivated by sourcing the `*_deactive.nu` scripts.

On Linux, the PATH can be changed during runtime by assigning to the PATH env var. So `let-env` and friends can be used for changing the PATH! However, this does not appear to work on Windows. I'll add the version that spawns a new shell in the next few days, so we have something working on Windows as well. I haven't tested either on MacOS.

A working `pathvar` command does solve part of the issue. However, I couldn't figure out an easy way to add many paths at once with `pathvar`. Everything that happens to the env in a function's scope does not transfer to the parent/global scope, so I'm not sure how to easily add several paths at once without resorting to setting `$nu.env.PATH` (which doesn't work on Windows? At least it appears not to on my Windows machine).

Thank you for the assistance and guidance on this issue!

**Comment by elferherrera** [comment]
@filaretov thanks a lot for the scripts.

So, if I understand correctly, `name-of-env` is the path to where the virtual env was created, right?
In your tests, do you think it is possible to load the environments by sourcing the activate.nu file?

I think it will help to have a file that can be sourced in a similar fashion to a bash script. This way we can start thinking about pushing these files to the `venv` python repo.  What do you think?

**Comment by filaretov** [comment]
Yep, that's how it works, I probably should have named the `venv` argument in a different way.

Generating a nu file to be sourced should be doable. I think that the PATH remains the most problematic right now, at least until 0.34.1 is released with a working `pathvar`. Specifically, the deactivation would be problematic on Windows, because I did not find a way to set the whole PATH easily. Python venvs save the current PATH before activating the env and then restore it upon deactivation. Since `let-env PATH = ...` didn't seem to work on Windows, there's no easy way to restore the PATH. Pathvar can be used to remove elements by index and in the case of venvs, only a single element is added, so `pathvar remove 0` should restore the old PATH. However, if a variable number of elements is added, I don't know how they could later be removed.

Should I open an issue describing the problem with using `let-env PATH = ...` on Windows? If I understand everything correctly, that seems to be the main blocker for implementing this in a cross-platform way. Once that is done, it would be feasible to integrate Nu into `venv`.

**Comment by fdncred** [comment]
@kubouch can you help with our `pathvar` understanding? is it working as it's supposed to or is there a bug in it in Windows? also, is there a way to add mutiple paths at once?

**Comment by elferherrera** [comment]
>Should I open an issue describing the problem with using let-env PATH = ... on Windows? If I understand everything correctly, that seems to be the main blocker for implementing this in a cross-platform way. Once that is done, it would be feasible to integrate Nu into venv.

I think it is a good idea. @kubouch is working on that and he will love the feed back

Regarding the script. Are you mostly working in linux or windows? if you are on linux, do you think you can sketch something that we can source so we can do some testing and improve on it?

**Comment by filaretov** [comment]
I work on both Windows and Linux regularly (which is why nushell is especially nice).

And yeah, I'd love to work on adding nu support in venv. I'll sketch an activate and deactivate script this week, and I'll take a loot at venv to see how a proper implementation should look like.

I'll also open an issue about the PATH on Windows.

I expect to have enough time to do it this week.

**Comment by elferherrera** [comment]
@filaretov I did a quick test with your file in a mac to load a python env
This is the file I sourced

```
def venv [name] {
	let venv-path = ($name | path expand)
	let venv-bin-path = ([$venv-path "bin"] | path join)
	let old-path = ($nu.path | str collect (path-sep))
	let new-path = ($nu.path | prepend $venv-bin-path | str collect (path-sep))
	[[name, value];
	 [PATH $new-path]
	 [VENV_OLD_PATH $old-path]
	 [VIRTUAL_ENV $venv-path]]
}

def windows? [] {
    (sys).host.name == "Windows"
}

def path-sep [] {
    if (windows?) { ";" } { ":" }
}

load-env (venv /Users/fjhe/Documents/playground/nushell/myproject/venv) <<<--- extra line
^^^^^^^ --> This line can be later changed automatically with python when creating the env
```
This worked nicely. I only had to source the file
```
source activate.nu
```

The only thing missing is changing the prompt so people can see that they are in a new env
@fdncred @jntrnr You once mentioned that the prompt could be changed via env variables. Am I correct?

**Comment by kubouch** [comment]
@fdncred `pathvar` should be fixed on Windows now. You cannot add multiple paths at once -- that's a limitation of `pathvar`. To add more paths you just need to call `pathvar add` multiple times and to remove them, you'd call `path remove 0` repeatedly. I think there is no fundamental reason why you shouldn't be able to pass, e.g., a list of paths to the `path add` command. It should be quite straightforward to implement.

@filaretov You cannot do `let-env PATH = ...` and you cannot define PATH with `load-env` either -- you need to use `pathvar` and/or set the `path` variable in config.toml. There is no `$nu.env.PATH` in Nushell, only `$nu.path`. The reason for this is that the path variable is platform-dependent (i.e., "Path" on Windows and "PATH" on Unix). Having it as an internal `$nu.path` we get the platform independence. Also, the `$nu.path` is stored as a list while the raw `Path`/`PATH` is a long string with platform-dependent separators. The `$nu.path` is then internally translated into `;`-separated `Path` (Windows) or `:`-separated `PATH` (Unix) environment variable, although I don't exactly remember how and when.

Honestly, I don't like how the path is special-cased from the rest of the environment variables. To unite them, though, we would have to come up with a system that solves the platform-independence and lets us define it as structured data while still keeping the serialized version for external commands. I think it's possible and I can write up some design thoughts about this on Discord but maybe that's out of scope for this issue.

**Comment by kubouch** [comment]
Well, apparently, you can set `PATH` with `let-env` and `load-env`, but you'll see the change under `$nu.path`, not `$nu.env.PATH`. Also, on Windows, it should be `Path`.

**Comment by elferherrera** [comment]
@kubouch do you think that it is possible to add a delete path to `pathvar` but instead of looking by index that it looks by name?
Or probably have a search subcommand that returns the index of the given path?

**Comment by elferherrera** [comment]
@kubouch Is pathvar supposed to delete all the paths?
![image](https://user-images.githubusercontent.com/6942205/127888638-5dcf26b0-7a85-4954-883c-203e75a6ef13.png)

**Comment by kubouch** [comment]
@elferherrera Removing by name should be definitely possible. If you check the `pathvar remove` subcommand, you'll see it collects the paths into `Vec<&str>` so it's just a matter of finding the path there.

@elferherrera Are you running the latest main? This looks like `pathvar` is still broken on your screenshot. I think the pathvar fix hasn't been released yet.

**Comment by elferherrera** [comment]
@kubouch Im running form the last release in mac. Let me compile in another computer to test the same

**Comment by kubouch** [comment]
@elferherrera I tried `pathvar remove 0` on a MacOS virtual machine from the latest main branch and it worked fine. Both `pathvar` and `$nu.path` showed with the first item removed. I'm wondering what's going on...

**Comment by fdncred** [comment]
@elferherrera the prompt is configurable via the `prompt = ` in the config.toml file. so, theoretically (i've never tried), if you had and env var, you could put it in the prompt. I make use of the `left_and_right_prompt.nu` in the nu_scripts repo, which sets up quite a few things in order to draw a prompt that looks like this.
![image](https://user-images.githubusercontent.com/343840/128021627-d668aa47-cb87-479a-a4e0-cc6af81ba78a.png)

**Comment by deeuu** [comment]
@elferherrera Thanks for the script, this appears to work well (I'm on Linux): I even get the virtual-env name in my [starship](https://starship.rs/) prompt. Are you planning on adding `deactivate` and submitting this to [nu_scripts](https://github.com/nushell/nu_scripts)? It would be lovely to have this auto-activate/-deactivate upon entering/leaving the project directory!

**Comment by fdncred** [comment]
Credit where credit is due. @filaretov has written the most recent activate scripts. I too hope they'll be PR'd into nu_scripts. The `autoenv` command might be able to be used to auto-activate these scripts upon entering a directory. Someone should try that because it would be so cool!

**Comment by filaretov** [comment]
Hey, thank you! Yup, I'll submit a PR tomorrow, I just wanted to use the scripts for a few days, as a precautionary measure.

**Comment by deeuu** [comment]
@filaretov Many thanks! @fdncred Perhaps I'll give the auto-activate a go, thanks for the pointer.

**Comment by elferherrera** [comment]
@deeuu I think the next step will be to test @filaretov 's script and when we are happy with it, it can be submitted to `venv` repository. This will allow python to generate the script when using `venv` module. I would assume something similar will be done with conda script. It's been a while since I used conda :P

Btw, this has been a blocker for many people. Thanks @filaretov for working on it.

**Comment by filaretov** [comment]
It was a pleasure, thanks to everyone for helping along the way.

In Conda's case, we have to define how Nushell creates and deletes environment variables, I haven't had a close enough look to see if it's going to work with Nushell's current features. One possible issue that might crop up is whether the `let-env` calls get called in the correct scope from Conda's `activate.py`. Other shells have ways of exporting env vars directly to the global scope (or rather, have no scope).

Since `venv` generates a script for every environment, the scripts themselves will be simpler, because they'll have hardcoded values for some things. They probably won't have to test for the current OS, for example. I'll be happy to look into it, once the current scripts have been tested a bit.

Also, I've completely ignored the prompt, because there wasn't much information on configuring it in the Nushell book and even there it links to Starship, which displays the virtual environment names based on the defined env vars. Maybe we could sidestep dealing with the prompt altogether, if most users rely on Starship anyway.

Another lesser issue is that we can't currently delete aliases or commands. Activated `venv` environments define a `deactivate` command, which "self-destructs" when calling it, so it doesn't exist when exiting the environment. In Nushell's case, we could just source one file for activating and another for deactivating, without creating an alias for it, so there's no pollution.

**Comment by fdncred** [comment]
great work @filaretov. thanks again! I'm closing this issue now. please reopen if needed.

**Comment by elferherrera** [comment]
>Another lesser issue is that we can't currently delete aliases or commands. Activated venv environments define a deactivate command, which "self-destructs" when calling it, so it doesn't exist when exiting the environment. In Nushell's case, we could just source one file for activating and another for deactivating, without creating an alias for it, so there's no pollution.

This is going to be a strange scenario for nushell, because the created function doesnt have access to the scope where the env variables were created. I think that for nushell's case we will need to have an activate and a deactivate script. I dont think there is another way around this

**Comment by iiwolf** [comment]
Nushell noob here who would really like to use Python and Nushell. I'm having trouble with these instructions.

1. The README says

> Source an activation script in your `config.toml`. For example, `conda.nu`.
   You'll then have the `conda-env` command available.

Where's this `config.toml` supposed to be? Is this a typo and it's meant to be `config.nu` instead?

2. Either way, when I try sourcing `conda.nu` I get an error about variable names. Are variables with dashes in them valid variable names?
``` 
nu_scripts\virtual_environments〉source conda.nu                                                             05/18/2023 03:38:15 PMError: nu::parser::parse_mismatch (link)

  × Parse mismatch during operation.
   ╭─[conda.nu:1:1]
 1 │ def conda-env [env-name] {
   ·                ────┬───
   ·                    ╰── expected valid variable name
 2 │     let conda-info = (conda info --envs --json | from json)
   ╰────
````

I suspect I'm misunderstanding multiple things. 

Thanks!

Isaac

**Comment by kubouch** [comment]
@iiwolf You can try one of the conda modules here https://github.com/nushell/nu_scripts/tree/main/modules/virtual_environments, there are also instructions in README.md. Modules are used with `use`, not `source`, see https://www.nushell.sh/book/modules.html for details about modules. Seems like you're using some old file.

`config.toml` does not exist anymore, we now have two config files `env.nu` and `config.nu` https://www.nushell.sh/book/configuration.html

**Comment by Chuong98CC** [comment]
> great work @filaretov. thanks again! I'm closing this issue now. please reopen if needed. For those interested, we've have these scripts on file at https://github.com/nushell/nu_scripts/tree/main/virtual_environments

@fdncred  The link is no longer working.

**Comment by fdncred** [comment]
https://github.com/nushell/nu_scripts/tree/main/modules/virtual_environments

**Comment by wangjiawen2013** [comment]
Hi,
`conda activate myenv` works well by this way, but we cannot exit the activated myenv after running `conda deactivate`. The `python` command will keep the same one, even though we have deactivated it!

Besides, we cannot run `conda deactivate` two times in nushell (we can run `conda deactivate` countless times in other shells such as bash ). It will show the following error at the second time:

~/work> conda deactivate
Error: nu::shell::column_not_found

  × Cannot find column 'CONDA_OLD_PATH'
    ╭─[/home/wangjw/work/nu_scripts-main/modules/virtual_environments/conda.nu:78:23]
 77 │     let path_name = if "PATH" in $env { "PATH" } else { "Path" }
 78 │     $env.$path_name = $env.CONDA_OLD_PATH
    ·                       ─────────┬─────────┬
    ·                                │         ╰── value originates here
    ·                                ╰── cannot find column 'CONDA_OLD_PATH'
 79 │ 
    ╰────
